Unit tests

pwilliams1 opened the pull request:
**Main Edits:**
- Added Test Classes for fsl_con_f, fsl_thr_clustfwep05, fsl_thr_voxelfwep05, spm_thr_clustunck10, spm_thr_voxelfdrp05, and spm_thr_voxelunct4
- Edited viewer.py to allow the main function to accept arguments (so we can run the viewer in the tests).
- Created a test to run the viewer against all available nidm files and check for errors (i.e program crashing).

**Note: A slight problem is that when the tests run, they create the html folders/files (because the tests run the viewer). At the moment I have to manually delete the files after running the tests.**

pwilliams1 committed:
Begin writing tests

pwilliams1 committed:
Start editing test for stats page

pwilliams1 committed:
Add SPM string test

pwilliams1 committed:
Test for FSL string

pwilliams1 committed:
Change file mode to x

pwilliams1 committed:
Add function to produce Z (Gaussianised T/F) string

pwilliams1 committed:
Fix problem with Z (Gaussianised T/F) not being displayed

pwilliams1 committed:
Fix gaussianised T/F problem for FSL case

pwilliams1 committed:
Add option to enter file to compare HTML page with

pwilliams1 committed:
Pull request updates

pwilliams1 committed:
Add test for statistic image type

pwilliams1 committed:
Change class name to fslStatsTests

pwilliams1 committed:
Add test for threshold value

pwilliams1 committed:
Add myString.split()

pwilliams1 committed:
Begin writing a general test for stats page

pwilliams1 committed:
Add test for fsl_thr_clustfwep05

pwilliams1 committed:
Add postats test

pwilliams1 committed:
Update threshold test

pwilliams1 committed:
Ensure correct iteration through file

pwilliams1 committed:
Add initial assignment of myString variable

pwilliams1 committed:
Add comment

pwilliams1 committed:
Edit general test

pwilliams1 committed:
Ensure myString variable is defined

pwilliams1 committed:
Begin adding comparison of version number and stat image

pwilliams1 committed:
Fix error with Conjunction Inference

pwilliams1 committed:
Change main() to have arguments (to enable tests to run it)

pwilliams1 committed:
Write test to run viewer.py on all nidm files required and check for errors

pwilliams1 committed:
Add Test folder

pwilliams1 committed:
Modify viewer.py so it returns destination (html) folder

pwilliams1 committed:
Add test_fsl_con_f.py to test viewer.py on nidm file

pwilliams1 committed:
Remove testError.py from main folder

pwilliams1 committed:
Add test for version number 6.00

pwilliams1 committed:
Add test for statistic image Z (Gaussianised T/F)

pwilliams1 committed:
Change self.file to self.postStatsFile to distinguish between files

pwilliams1 committed:
Add class for fsl_thr_clustfwep05

pwilliams1 committed:
Add Tests for cluster threshold etc for fsl_thr_clustfwep05

pwilliams1 committed:
Add comments

pwilliams1 committed:
Add class for fsl_thr_voxelfwep05 tests

pwilliams1 committed:
Add class for fsl_thr_voxelfwep05 tests

pwilliams1 committed:
Add voxel wise tests

pwilliams1 committed:
Add test for spm_thr_clustunck10

pwilliams1 committed:
Add spm_thr_voxelfdrp05 test and change myString to self.myString

pwilliams1 committed:
Add test for ex_spm_thr_voxelunct4

pwilliams1 committed:
Rename test_fsl_con_f.py to test_fslViewer.py

pwilliams1 committed:
Remove redundant comment line

pwilliams1 committed:
Fix pull request error

pwilliams1 committed:
Fix pull request error

pwilliams1 committed:
Remove if statements in main function - Fixes indented block error

cmaumet commented:
@pwilliams1: one quick comment: can you move all the test files into the `Tests` folder? It looks like a couple of them are still at the root of the repo.

And also, if you can find this (brand new!) option in your repo (https://github.com/blog/2247-improving-collaboration-with-forks), please activate it so that I can contribute as part of my review of this PR.

pwilliams1 committed:
Remove redundant tests

cmaumet commented:
Can you also add a `Test` title in the `README.md` and include the actual commands you use to run the tests.

pwilliams1 committed:
Update README to include tests

pwilliams1 commented:
I have deleted the other tests as they are now contained in one script inside the Tests folder. I have also updated the README.

Sorry, I can't find this option, there doesn't seem to be a checkbox available for me to select.

Peter

---

From: Camille Maumet notifications@github.com
Sent: 08 September 2016 16:09:58
To: pwilliams1/nidmresults-fslhtml
Cc: Williams, Peter; Mention
Subject: Re: [pwilliams1/nidmresults-fslhtml] Unit tests (#10)

@pwilliams1https://github.com/pwilliams1: one quick comment: can you move all the test files into the Tests folder? It looks like a couple of them are still at the root of the repo.

And also, if you can find this (brand new!) option in your repo (https://github.com/blog/2247-improving-collaboration-with-forks), please activate it so that I can contribute as part of my review of this PR.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/pwilliams1/nidmresults-fslhtml/pull/10#issuecomment-245630001, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ATdQEdBnb2LD4_DMbF-ut3CRuhR-BJuOks5qoCVGgaJpZM4J3_sb.

cmaumet commented:
Great thanks! (Maybe the feature is only available in PR sent to another repository). I will send you my comments directly in this PR.

pwilliams1 commented:
No problem.

cmaumet commented:
@pwilliams1: could you go through the following suggestions?
- To invoke the tests, the `README.md` should really say:
  
  ```
  python Tests/test_fslViewer.py
  ```
  
  (we assume that the user is at the root of the project). If you prefer not to have the `python` prefix that is fine too but then we should add `#!/usr/bin/env python` at the top of the `test_fslViewer.py` file (and similarly for other tests) to let the program know this is python (cf. http://stackoverflow.com/questions/15587877/run-a-python-script-in-terminal-without-the-python-command for more info). If in Windows you might not need the `python` prefix but my guess is that it is because the code is executed from a python console.
- The `main` function is missing in `test_fslViewer.py`. 
- Please tick the tests you have done from the list in https://github.com/pwilliams1/nidmresults-fslhtml/issues/5. If everything in this issue is solved, then you can add a comment linking to this pull request and close the issue.
- Can you move the test data into a `data` folder (inside `Tests`) and make sure the test script still work.

pwilliams1 committed:
Add #!/usr/bin/env python3 to top of test scripts

pwilliams1 committed:
Add main function to test_fslViewer.py

pwilliams1 committed:
Change description of how to run tests

pwilliams1 commented:
Regarding the first bullet point:

I've realised that running the tests one-by-one with Tests/testname.py won't work, as the tests have a line import viewer, and viewer.py is not in the Tests folder. I have been running the tests using python -m unittest discover Tests, which works when you run it from the root of the project. Do we want to put viewer.py inside the Tests folder?

---

From: Camille Maumet notifications@github.com
Sent: 09 September 2016 12:31:35
To: pwilliams1/nidmresults-fslhtml
Cc: Williams, Peter; Mention
Subject: Re: [pwilliams1/nidmresults-fslhtml] Unit tests (#10)

@pwilliams1https://github.com/pwilliams1: could you go through the following suggestions?
-   To invoke the tests, the README.md should really say:

python Tests/test_fslViewer.py

(we assure that the user is at the root of the project). If you prefer not to have the python prefix that is fine too but then we should add #!/usr/bin/env python at the top of the test_fslViewer.py file (and similarly for other tests) to let the program know this is python (cf. http://stackoverflow.com/questions/15587877/run-a-python-script-in-terminal-without-the-python-command for more info). If in Windows you might not need the python prefix but my guess is that it is because the code is executed from a python console.
-   The main function is missing in test_fslViewer.py.
-   Please tick the tests you have done from the list in #5https://github.com/pwilliams1/nidmresults-fslhtml/issues/5. If everything in this issue is solved, then you can add a comment linking to this pull request and close the issue.
-   Can you move the test data into a data folder (inside Tests) and make sure the test script still work.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/pwilliams1/nidmresults-fslhtml/pull/10#issuecomment-245888645, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ATdQEfZzhSmgPlWMJU9dw22PZ0hbiaOFks5qoUOXgaJpZM4J3_sb.

cmaumet commented:
We do want to keep the tests in a separate folder from the main code (i.e. the viewer). The tests can be run with the commands you provided as long as you add the folder containing `viewer.py` to your python path. On Unix systems, that can be done with:

```
PYTHONPATH="<PATH_TO_THE_VIEWER>:$PYTHONPATH"
export PYTHONPATH
```

(where `<PATH_TO_THE_VIEWER>` is replaced by the actual path). On Windows, I found this: http://stackoverflow.com/questions/3701646/how-to-add-to-the-pythonpath-in-windows-7. 

Let me know if you manage to get the tests running again.

cmaumet commented:
But you might need Admin rights... What program do you use to run your python code on Windows?

pwilliams1 commented:
I do need admin rights to access systems settings.

For running my code, I use the command line and just type the name of the script (e.g. viewer.py).

---

From: Camille Maumet notifications@github.com
Sent: 09 September 2016 14:34:36
To: pwilliams1/nidmresults-fslhtml
Cc: Williams, Peter; Mention
Subject: Re: [pwilliams1/nidmresults-fslhtml] Unit tests (#10)

We do want to keep the tests in a separate folder from the main code (i.e. the viewer). The tests can be run with the commands you provided as long as you add the folder containing viewer.py to your python path. On Unix systems, that can be done with:

PYTHONPATH="<PATH_TO_THE_VIEWER>:$PYTHONPATH"
export PYTHONPATH

(where <PATH_TO_THE_VIEWER> is replaced by the actual path). On Windows, I found this: http://stackoverflow.com/questions/3701646/how-to-add-to-the-pythonpath-in-windows-7.

Let me know if you manage to get the tests running again.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/pwilliams1/nidmresults-fslhtml/pull/10#issuecomment-245914290, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ATdQEW4jZ1NQK9PYT6xEhRF5yTle1v0Pks5qoWBsgaJpZM4J3_sb.

cmaumet commented:
It looks like you can also set your path temporarily from the command line, cf. https://docs.python.org/3/using/windows.html#configuring-python. Let me know if that works.

pwilliams1 committed:
Change glob section to make glob look in data folder

pwilliams1 commented:
Thanks, that worked! Will now check that the tests work after including a data folder.

pwilliams1 committed:
Begin fixing problem with setUp running at the start of every test method

pwilliams1 committed:
Begin fixing setUp problem (setUp runs at the start of every test)

pwilliams1 committed:
Edit test_fslViewer so that viewer is run only once (on each data file)

pwilliams1 committed:
Change setUp method to only open the postStats file (not create it, this now takes place outside of the tests)

pwilliams1 committed:
Begin changing other setUp functions for all tests

pwilliams1 committed:
Begin using os.path features to access data folder

pwilliams1 committed:
Change code to run viewer on every turtle file in data folder (using os.path features and glob)

cmaumet commented on Tests/test_fslViewer.py:
```diff
@@ -0,0 +1,314 @@
+	
+	def setUp(self): #Open necessary file
+		self.myString = ""
+		self.data = "data\\fsl_con_f.nidm.ttlTestResults\\postStats.html"
```
@pwilliams1: could you use `os.path.join` as described [here](https://docs.python.org/2/library/os.path.html#os.path.join) rather than backslashes? This will ensure that the code works with different OS (Windows, Unix...). This would be similar to what you have already done in 6425bb3.

pwilliams1 replied:
No problem, changing it now.

pwilliams1 committed:
Use os.path.join instead of double backslash, and change threshold test (P = vs just P)

pwilliams1 committed:
Add seperate test for threshold value (e.g. P = 0.05)

pwilliams1 committed:
Remove redundant print statements

pwilliams1 committed:
Change software name test to check for both FMRI and FSL/SPM in file (avoids clash with title FSL Viewer)

pwilliams1 commented:
It should be working now - Have run the tests using a 'data' folder and they seem to work.

pwilliams1 committed:
Remove shutil module usage

pwilliams1 committed:
Use os.path.join and remove redundant print statement

pwilliams1 committed:
Add comments

pwilliams1 commented:
The tests work, but when running testError.py I get a **ResourceWarning: unclosed file**... Not sure what is causing this.

cmaumet commented:
Thank you!

Is the warning message saying `ResourceWarning: unclosed file`? If so, it might be caused by the `rdflib` library: I have reported such warning in https://github.com/RDFLib/rdflib/issues/651. But it could also be that you have open (with the `open` function) a file but not closed it (with the `close` function). Can you double-check?

I am not going to run the tests on my machine as I would rather wait until we have set the tests to automatically download data from NeuroVault. So, can you just paste in: the command you run and the output you get in a comment below?

Once this is done, I will merge this PR.

pwilliams1 commented:
Yes it says that. I do not get this problem when running viewer.py outside of the tests, and testError.py does not open any files apart from when calling viewer.py

Sorry, do you want the output of the unclosed file error or the output of the tests (pass/fail etc)?

---

From: Camille Maumet notifications@github.com
Sent: 13 September 2016 14:01:37
To: pwilliams1/nidmresults-fslhtml
Cc: Williams, Peter; Mention
Subject: Re: [pwilliams1/nidmresults-fslhtml] Unit tests (#10)

Thank you!

Is the warning message saying ResourceWarning: unclosed file? If so, it might be caused by the rdflib library: I have reported such warning in RDFLib/rdflib#651https://github.com/RDFLib/rdflib/issues/651. But it could also be that you have open (with the open function) a file but not closed it (with the close function). Can you double-check?

I am not going to run the tests on my machine as I would rather wait until we have set the tests to automatically download data from NeuroVault. So, can you just paste in: the command you run and the output you get in a comment below.

Once this is done, I will merge this PR.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/pwilliams1/nidmresults-fslhtml/pull/10#issuecomment-246673562, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ATdQEe2y51V73l8k9sEX1naxguqJPvYhks5qpp6xgaJpZM4J3_sb.

cmaumet commented:
Yes, the warning is only displayed in test mode and otherwise silent, which is why you don't see it when calling the viewer directly. 

Sorry, I meant pasting the command and output of the tests. (In the future, the tests report will be available on Travis but for now a comment will be fine).

pwilliams1 commented:
Command: Tests\test_fslViewer.py

[cid:2dcfc193-86f5-403e-95b0-30e119c45538]


Command: Tests\testError.py


[cid:38cdc7f6-1eb4-49f4-b431-c1b5fcf907b2]



________________________________
From: Camille Maumet <notifications@github.com>
Sent: 13 September 2016 14:17:12
To: pwilliams1/nidmresults-fslhtml
Cc: Williams, Peter; Mention
Subject: Re: [pwilliams1/nidmresults-fslhtml] Unit tests (#10)


Yes, the warning is only displayed in test mode and otherwise silent, which is why you don't see it when calling the viewer directly.

Sorry, I meant pasting the command and output of the tests. (In the future, the tests report will be available on Travis but for now a comment will be fine).

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pwilliams1/nidmresults-fslhtml/pull/10#issuecomment-246677574>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATdQEXXDJ_6YaUvOLNTP4u2bb5FRsSACks5qpqJYgaJpZM4J3_sb>.

pwilliams1 commented:
Command: `Tests\testError.py`

![testerroroutput](https://cloud.githubusercontent.com/assets/20402193/18475285/c4af8438-79bd-11e6-8fa3-2bd5011c6a93.png)

Command: `Tests\test_fslViewer.py`

![test_fslvieweroutput](https://cloud.githubusercontent.com/assets/20402193/18475312/dbb32464-79bd-11e6-8623-5a3015c535d4.png)

cmaumet commented:
Excellent! Thanks! Let's merge!

Can you make sure to use a fixed-width font to outline commands/code or terminal outputs you include in a comment? This is a convention and it makes the comments easier to read. For that, you can use the "`" character in markdown (cf. "inline code" in https://guides.github.com/features/mastering-markdown/). I have updated your comment above.

cmaumet commented:
Hi @pwilliams1! Now that this PR is merged, you will have to update your local copy of the `master` branch with:

```
git checkout master
git pull origin master
```

And then you can create a new branch from here (replacing `<MY_NEW_BRANCH>` with the actual name):

```
git checkout -b <MY_NEW_BRANCH>
```

 or, if you were already working on another branch, you can merge the changes made to master with (replacing `<MY_OTHER_BRANCH>` with the actual name):

```
git checkout <MY_OTHER_BRANCH>
git merge master
```

pwilliams1 commented:
New branch created - I assume I should start working on changing the viewer to use **dominate** now?

cmaumet commented:
Yes, that would be great!

By the way, you can also delete the (now merged) `unit_tests` branch on your machine with:

```
git branch -d unit_tests
```

(Note that it is important to use a lower case `d`, not upper-case).
